### PR TITLE
Add T20 ruff rule (print calls)

### PR DESCRIPTION
**Pull request by brisvag**
# References and relevant issues
Tracking issue: #5589

Previous discussion: #6840

# Description

As asked in #6840, I split out this more controversial rule from it, so we can discuss it separately.

### CODE REVIEW

**Comment by brisvag** [comment]
@jni input appreciated!

**Review by Czaki — APPROVED** [review]

File: napari/_tests/test_import_time.py [hunk 1]
```diff
@@ -19,7 +19,8 @@ def test_import_time(tmp_path):
     # # figure out how to use it
     # assert name == 'napari'
     # assert int(time) < 1_000_000, "napari import taking longer than 1 sec!"
-    print(f'\nnapari took {int(time)/1e6:0.3f} seconds to import')
+    # print(f'\nnapari took {int(time)/1e6:0.3f} seconds to import')
+    # TODO: can we move this to benchmarks?
```
Comment (jni) [inline napari/_tests/test_import_time.py]: do we want to do this before merging?

**Comment by jni** [comment]
This PR changes 11 prints to print # noqa, and 7 prints to logging/warn. (And note that for our purposes logging is in fact equivalent to print.) imho, that is a bit of a smell that the rule is too stringent and we should not activate it. I'm not actively going to block it but yeah, I don't love it.

File: napari/utils/notifications.py [hunk 2]
```diff
@@ -397,9 +397,9 @@ def show_console_notification(notification: Notification):
         ):
             return
 
-        print(notification)
+        print(notification)  # noqa: T201
     except Exception:
-        print(
+        print(  # noqa: T201
```
Comment (Czaki) [inline napari/utils/notifications.py]: I prefer to change this to [`logging.exception`](https://docs.python.org/3.8/library/logging.html#logging.exception)

File: napari/_qt/dialogs/qt_notification.py [hunk 3]
```diff
@@ -405,9 +405,11 @@ def _debug_tb(tb):
     QApplication.processEvents()
     QApplication.processEvents()
     with event_hook_removed():
-        print("Entering debugger. Type 'q' to return to napari.\n")
+        print(  # noqa: T201
```
Comment (Czaki) [inline napari/_qt/dialogs/qt_notification.py:408]: `loggin.info`?

File: napari/_tests/test_import_time.py (hunk 1 above)
Comment (Czaki) [reply napari/_tests/test_import_time.py]: It may be worth

**Comment by Czaki** [comment]
> (And note that for our purposes logging is in fact equivalent to print.)

And we should change this. We should implement a custom logger grabber and GUI to inspect logs from application.

**Comment by jni** [comment]
> And we should change this. We should implement a custom logger grabber and GUI to inspect logs from application.

Ok I'm convinced. Maybe changing the remaining prints to logs then makes sense.

**Comment by Czaki** [comment]
> Agree it's not a super-important rule, but I think it does more good than bad. Agree we should have a log handler we grab everywhere! It could be done as part of this PR as well. @Czaki I'm not sure I follow the GUI comment; can you elaborate?

Something like [icy](https://icy.bioimageanalysis.org/) Output tab:
obraz

Or console log in ImageJ. So place where it will be accessible to user who launch napari without a terminal.

It may be even smarter (like runtime level filtering, or selection of which logger output should be published).

**Comment by jni** [comment]
I don't think the logging console needs to be part of this PR (indeed, maybe not even this repo 😂), but rather, change every remaining print to a log. What do you think @brisvag @Czaki ?

**Comment by Czaki** [comment]
> I don't think the logging console needs to be part of this PR (indeed, maybe not even this repo 😂)

I think it should be part of the default installation `napari[all]`, but maybe outside this repo. But if @brisvag want, it could be even part of this PR. 

> but rather, change every remaining print to a log.

No. Output of `napari --info` should be printed using print, not logger.

**Comment by brisvag** [comment]
> I think it should be part of the default installation `napari[all]`, but maybe outside this repo. But if @brisvag want, it could be even part of this PR.


Definitely not this PR, but I agree that it would be nice to just have it somewhere. As long as we use logging properly, we should be able to do this at a later stage. 

> No. Output of `napari --info` should be printed using print, not logger.

Yeah, and a couple other things!

@Czaki: what's the general idea behing using `warnings.warn` vs `logging.warning`? We seem to use them randomly, maybe we should always use log?

**Comment by Czaki** [comment]
> @Czaki: what's the general idea behing using `warnings.warn` vs `logging.warning`? We seem to use them randomly, maybe we should always use log?

warnings are shown as modal dialog in the bottom-right corner and logging is not displayed. 

I think that warning is something that should be immediately displayed to the user and logging for later access.

**Comment by brisvag** [comment]
But don't we want the warnings to the user to also be displayed in the logs? Wouldn't it be cleaner to take all logs from level `WARN` or more and push them as notifications, instead of using two different systems? I also just confused about why python offers these two systems in the standard library...

**Comment by Czaki** [comment]
For example, warning by default provides deduplication of events (with well, known bug). And logging push to logs every event that happens. 

Warnings were introduced in python 2.1 and logging in python 2.3 (based on documentation). 

Warnings are global, logging offers the option to have multiple loggers with different settings. 

I see that there are similar, but not equivalent. I'm not sure if we want to change every logging into notification.

**Comment by brisvag** [comment]
I changed 1 more print into log, but I think the rest should remain `noqa` because they are all special cases. Let me know if you disagree!

I also moved the import time to a benchmark based on the implementation at https://github.com/proost/pandas/blob/master/asv_bench/benchmarks/package.py. Let's see if it works :)

File: napari/_tests/test_import_time.py (hunk 1 above)
Comment (brisvag) [reply napari/_tests/test_import_time.py]: done!

File: napari/utils/notifications.py (hunk 2 above)
Comment (Czaki) [reply napari/utils/notifications.py]: @brisvag and this?

**Comment by Czaki** [comment]
@brisvag It looks like the last merge to main introduced merge conflict

File: napari/utils/notifications.py (hunk 2 above)
Comment (brisvag) [reply napari/utils/notifications.py]: isn't this a place where we want the user to see it? So maybe `warnings`?

File: napari/_qt/dialogs/qt_notification.py (hunk 3 above)
Comment (brisvag) [reply napari/_qt/dialogs/qt_notification.py:408]: Not sure how to deals with these cause I'm not familiar with this module. Isn't this print here on purpose so the traceback can be shown to the user?

File: napari/utils/notifications.py (hunk 2 above)
Comment (Czaki) [reply napari/utils/notifications.py]: Looking at this code, I think that warning may not be rendered.

Comment (brisvag) [reply napari/utils/notifications.py]: But warnings should show up as popups right?

Comment (Czaki) [reply napari/utils/notifications.py]: yes. But see code above. There is only settings interaction. So if interaction with settings object fail, you expect GUI to be working?

Comment (brisvag) [reply napari/utils/notifications.py]: I guess no :P but a `warnings.warn` will be rendered on the terminal as well right?

Comment (Czaki) [reply napari/utils/notifications.py]: Lets do this this way

Comment (Czaki) [reply napari/utils/notifications.py]: Or maybe not. The `logging.exception` will also print exception information. There is no equivalent for warnings. If you prefer warning, then we should print stacktrace somehow.

Comment (brisvag) [reply napari/utils/notifications.py]: ok, logging is fine by me then. We should really get a custom handler going afterwards :)

Comment (brisvag) [reply napari/utils/notifications.py]: but do we still raise after? I guess so?

Comment (Czaki) [reply napari/utils/notifications.py]: See comment above raise.

**Review by Czaki — APPROVED** [review]

File: napari/utils/perf/_patcher.py [hunk 4]
```diff
@@ -183,18 +184,18 @@ def my_announcer(wrapped, instance, args, kwargs):
...
             module, attribute_str = _import_module(target_str)
             if module is not None and attribute_str is not None:
                 _patch_attribute(module, attribute_str, patch_func)
-        except PatchError as exc:
+        except PatchError:
             # We don't stop on error because if you switch around branches
             # but keep the same config file, it's easy for your config
             # file to contain targets that aren't in the code.
-            print(f'Patcher: [ERROR] {exc}')
+            logging.exception('Something went wrong while patching')
```
Comment (jni) [inline napari/utils/perf/_patcher.py:201]: Should we not log the exception message itself?

**Review by jni — APPROVED** [review]

File: napari/utils/perf/_patcher.py (hunk 4 above)
Comment (brisvag) [reply napari/utils/perf/_patcher.py:201]: `logging.exception` logs the exception without passing it explicitly by doing some magic inspection of the stack trace!

Comment (jni) [reply napari/utils/perf/_patcher.py:201]: 🤯 

Can you add a comment above it pointing that out? Cos that is just batshit crazy. 😂

**Comment by jni** [comment]
LOL @brisvag literally the day after you merge this, with much resistance from me:

```
~/projects/napari on monitor-fps(✗) at 0:00:15
$ git commit -a -m "Fix docstring indentation"
ruff-format..............................................................Passed
ruff.....................................................................Failed
- hook id: ruff
- exit code: 1

napari/_qt/qt_viewer.py:1070:13: T201 `print` found
napari/_qt/qt_viewer.py:1074:13: T201 `print` found
napari/_qt/qt_viewer.py:1080:9: T201 `print` found
Found 3 errors.
```

😅

**Pull request merged**